### DISCUSSION TITLE
Cache size and eviction

### DISCUSSION THREAD

**Discussion by Poongavanama** [ben-manes/caffeine#1071] [Q&A] [answered]
Can you please clarifiy the below points.

1.Cache eviction vs cache expiry which takes precedence.
2.Size is it entries vs MB.
3. How does cache weighted eviction works?
4. Eviction is invalidation() or delete() process

Thank you.

**Comment by ben-manes** [comment]
1. The cache will remove expired entries before it performs size eviction.
2. This can be adjusted by using a [weigher](https://github.com/ben-manes/caffeine/wiki/Eviction#size-based). The default is that every entry takes 1 unit of capacity (entry count). The weigher could instead calculate the entry's byte size as the number of units, e.g. by using [JAMM](https://github.com/jbellis/jamm) or [JOL](http://openjdk.java.net/projects/code-tools/jol/). The byte size is discussed in [JEP draft: Low-level Object layout introspection methods](https://openjdk.org/jeps/8249196) with links to Solr and Cassandra examples which use a Caffeine weigher to achieve this.
3. The weight is not used to make eviction decisions and is simply a threshold value. There was a [research paper](https://arxiv.org/pdf/2105.08770.pdf) which goes into more detail, but that work hasn't been incorporated yet. 
4. We [define](https://github.com/ben-manes/caffeine/wiki/Removal) the terms as below and when notifying a listener we send along the [RemovalCause](https://www.javadoc.io/doc/com.github.ben-manes.caffeine/caffeine/latest/com.github.benmanes.caffeine/com/github/benmanes/caffeine/cache/RemovalCause.html).  
   1. **eviction** means removal due to the policy
   2. **invalidation** means manual removal by the caller
   3. **removal** occurs as a consequence of invalidation or eviction

**Comment by Poongavanama** [comment]
Thank you @ben-manes

**Comment by Poongavanama** [comment]
Whcih one to recommented to use in configuration bean expireAfterWrite or expireAfterAccess, We have kubernetis nodes and multiple request invoke the API. Could you please suggesst which one is recommented to use.
Currently I am suing expireAfterAccess.

**Reply by ben-manes** [reply, marked as answer]
I don’t know, they are different policies, time-to-idle or time-to-live. I think TTL (expireAfterWrite) is more common, but it depends on what behavior you want.

**Reply by Poongavanama** [reply]
Thank you very much @ben-manes.

**Comment by Poongavanama** [comment]
> I don’t know, they are different policies, time-to-idle or time-to-live. I think TTL (expireAfterWrite) is more common, but it depends on what behavior you want.
